core/os/memory.cpp

```diff
@@ -142,21 +171,25 @@ void Memory::free_static(void *p_ptr, bool p_pad_align) {
 	bool prepad = true;
 #else
 	bool prepad = p_pad_align;
+	alloc_count--; // no mutex for performance
 #endif
 
-	alloc_count--;
-
 	if (prepad) {
 		mem -= PAD_ALIGN;
-		uint64_t *s = (uint64_t *)mem;
-
+		const uint64_t s = *((uint64_t *)mem);
 #ifdef DEBUG_ENABLED
-		mem_usage -= *s;
+		STATS_ENTER_CRIT_SECT
+		mem_usage -= s;
+		alloc_count--;
+		STATS_EXIT_CRIT_SECT
+		// Ease debugging of dangling pointers
+		memset(mem, 0xEA, PAD_ALIGN + s);
 #endif
-
 		free(mem);
 	} else {
-
+#if !defined DEBUG_ENABLED
+#error "Debug builds expect prepadding to be enabled."
```

Comment:
I feel this is going to error out on all non-debug builds, as it is a preprocessor directive.